File: go/excercise415/add_strings.go
package excercise415

import (
	"strconv"
	"strings"
)

func addStrings(s1, s2 string) string {
	i := len(s1) - 1
	j := len(s2) - 1
	carry := 0

	result := make([]string, 0, maxInt(i, j)+1)
	for i >= 0 || j >= 0 || carry != 0 {
		x, y := 0, 0

		if i >= 0 {
			x = int(s1[i] - '0')
			i--
		}

		if j >= 0 {
			y = int(s2[j] - '0')
			j--
		}

		sum := x + y + carry
		result = append([]string{strconv.Itoa(sum % 10)}, result...)
		carry = sum / 10
	}

	return strings.Join(result, "")
}

func maxInt(n1, n2 int) int {
	if n1 > n2 {
		return n1
	}
	return n2
}

File: python/excercise415/add_strings.py
def add_strings(s1: str, s2: str) -> str:
    i, j = len(s1)-1, len(s2)-1
    carry = 0
    result = []

    while i >= 0 or j >= 0 or carry != 0:
        digit1 = int(s1[i]) if i >= 0 else 0
        digit2 = int(s2[j]) if j >= 0 else 0

        total = digit1 + digit2 + carry
        result.append(str(total % 10))
        carry = total // 10

        i -= 1
        j -= 1

    return ''.join(result[::-1])
